sui_programmability/examples/defi/sources/pool.move

```diff
@@ -0,0 +1,309 @@
+
+        // TODO: Figure out what to do with this buddy.
+        // let coin2_added = (share_minted * tok_amount) / lsp_supply;
+
+        Balance::join(&mut pool.sui, sui_balance);
+        Balance::join(&mut pool.token, token_balance);
+
+        Transfer::transfer(
+            Coin::mint(share_minted, &mut pool.lsp_treasury, ctx),
+            TxContext::sender(ctx)
+        );
+    }
+
+    /// Remove liquidity from the `Pool` by burning `Coin<LSP>`.
+    /// Sender gets `Coin<T>` and `Coin<SUI>`.
+    public(script) fun remove_liquidity<T>(
+        pool: &mut Pool<T>,
+        lsp: Coin<LSP<T>>,
+        ctx: &mut TxContext
+    ) {
+        let lsp_amount = Coin::burn(lsp, &mut pool.lsp_treasury);
+        let (sui_amt, tok_amt, lsp_supply) = get_amounts(pool);
+
+        let sui_removed = (sui_amt * lsp_amount) / lsp_supply;
+        let tok_removed = (tok_amt * lsp_amount) / lsp_supply;
+
+        let sender = TxContext::sender(ctx);
+
+        Transfer::transfer(
+            Coin::withdraw(&mut pool.sui, sui_removed, ctx),
+            sender
+        );
+
+        Transfer::transfer(
+            Coin::withdraw(&mut pool.token, tok_removed, ctx),
+            sender
+        );
+    }
+
+    /// Get most used values in a handy way:
+    /// - amount of SUI
+    /// - amount of token
+    /// - total supply of LSP
+    public fun get_amounts<T>(pool: &Pool<T>): (u64, u64, u64) {
+        (
+            Balance::value(&pool.sui),
+            Balance::value(&pool.token),
+            Coin::total_supply(&pool.lsp_treasury)
+        )
+    }
+
+    /// Calculate the output amount minus the fee - 0.3%
+    fun get_input_price(
+        input_amount: u64, input_reserve: u64, output_reserve: u64
+    ): u64 {
+        let input_amount_with_fee = input_amount * 997; // 0.3% fee
+        let numerator = input_amount_with_fee * output_reserve;
+        let denominator = (input_reserve * 1000) + input_amount_with_fee;
+
+        numerator / denominator
+    }
+
+    #[test_only]
+    public fun get_price_for_sui<T>(pool: &Pool<T>, to_sell: u64): u64 {
+        let (sui_amt, tok_amt, _) = get_amounts(pool);
+        get_input_price(to_sell, tok_amt, sui_amt)
+    }
+
+    #[test_only]
+    public fun get_price_for_token<T>(pool: &Pool<T>, to_sell: u64): u64 {
+        let (sui_amt, tok_amt, _) = get_amounts(pool);
+        get_input_price(to_sell, sui_amt, tok_amt)
+    }
+
+    #[test_only]
+    public fun init_for_testing(ctx: &mut TxContext) {
+        init(ctx)
+    }
+}
+
+#[test_only]
+module defi::pool_tests {
+    use sui::SUI::SUI;
+    // use sui::Balance::{Self};
+    use sui::Coin::{Self, Coin, mint_for_testing as mint, };
+    use sui::TestScenario::{Self as test, Scenario, next_tx, ctx};
+    use defi::pool::{Self, Pool, LSP};
+
+    // Gonna be our test token.
+    struct BEEP {}
+
+    // Tests section
+    #[test] public(script) fun test_init_pool() { test_init_pool_(&mut scenario()) }
```

Comment:
Do these need to be `public(script)` (similar ? for `test_init_pool`)?

Reply:
They need to be `public(script)` to call `tests_` that are also `public(script)` which call pool's `public(script)` methods. 😕

Reply:
soon